### PR TITLE
Support Scala.js 1.5.0

### PR DESCRIPTION
**Pull request by clutroth**
I added scalajs 1.5.0 support. A allowed using deprecated "file as output" with warring, and ported to new API. WDYT?

### CODE REVIEW

File: .gitignore
Comment (gzm0) [inline .gitignore]: We usually don't put IDE specific things into the checked-in `.gitignore`. Put it in your global `.gitignore`?

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: I'm not sure this is very user friendly. Essentially, if the output directory doesn't exist, instead of failing or creating it, it falls back to the old behavior (potentially creating a file where a directory is expected).

I think another flag that is mutually exclusive with this one is needed. Unfortunately, the obvious candidate `-d` (same option as in `scalac` / `javac`) is already taken. Not sure what a good flag would be.

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala:95]: There were two additional configuration options introduced in Scala.js 1.5.0:
- `moduleSplitStyle`
- `outputPatterns`

We might want to add these as well.

File: build.sbt
Comment (gzm0) [inline build.sbt]: We need a better solution for this :) Have a look here (not pretty, but the best way we could find):

https://github.com/scala-js/scala-js/blob/d09ce384d35e3310272e8099dda5f2d3fee8dac1/sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala#L265-L292

File: scripts/test-cli.sh
Comment (gzm0) [inline scripts/test-cli.sh]: It might be nice to use `SplitStyle.SmallestModules` here to make sure the wiring is actually correct.

File: scripts/test-cli.sh
Comment (gzm0) [inline scripts/test-cli.sh]: Please put a newline at the end of the file.

**Review by gzm0 — CHANGES_REQUESTED** [review]
Thank you for opening this PR!

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (clutroth) [reply src/main/scala/org/scalajs/cli/Scalajsld.scala]: Ok, so I'll restore exact previous behavior, and introduce new option. IMHO sustaining backward compatibility is more important than one-letter option. I'll introduce `output-dir` option for directory output, and forbid simultaneous use of `-o`, and - `output-dir` . WDYT?

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala:80]: I think the `OutputFile` classes are not necessary anymore now.

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: call this `jsFilePattern`? I think for now it is OK to only support this specific type of pattern, but we should reserve the `outputPatterns` name for that option, if we ever want to introduce it.

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: Module splitting *style* (style and strategy really both don't add any semantic value here IMO, but then we should stick with the terminology that already exists).

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: Space after `:`.

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: I'm not a fan of not having a short option for a de-facto required flag (it's only not required for deprecated usage).

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: `<dir>`?

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: I think this help text can be improved. Currently, it describes what the flag does from the perspective of the CLI implementation. It should describe what the flag does from the perspective of the CLI user (i.e. change the names of the files that are written).

Further, it should point to documentation explaining what a valid pattern / value is (other texts might also fall short of this, depending on how much scopt auto-generates; feel free to improve that in a separate commit / PR).

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: Have you considered (both different flow and improved error messages).

```scala
if (c.output.isDefined)
  reportWarning("using a single file as output (--output) is deprecated since Scala.js 1.3.0. Use --outputDirectory instead.")

if (c.outputDir.isDefined == c.output.isDefined)
  failure("exactly one of --output or --outputDir have to be defined")
else
  success
```

Note that this warns on --output, even if --outputDir is defined. IMHO this is a good thing: If both are set, we nudge the user into the right direction right away. They wont remove --outputDir, just to find that --output is deprecated.

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: Why this change? (seems unrelated).

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: I think reading the arguments here directly to decide which overload to call is fine.

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: Spaces around `{`.

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: Add a comment? The name of the class is a bit meta IMO. (can stay, but should be commented on).

Maybe: DeprecatedLinkerAPI is a more descriptive name?

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: Just type-ascribe the object at call site? It's only used once, right?

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: Indentation here is not Scala.js style. Have a look here: https://github.com/scala-js/scala-js/blob/master/CODINGSTYLE.md#indentation

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: You don't need to pass the ec explicitly.

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: It might be OK to inline this in the method above.

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: Remove?

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: Remove?

File: scripts/test-cli.sh
Comment (gzm0) [inline scripts/test-cli.sh]: These file names are not specified (or more precisely: the module ID is an implementation detail subject to change). Maybe just assert that there is more than one `.js` file?

File: scripts/test-cli.sh
Comment (gzm0) [inline scripts/test-cli.sh]: It's maybe a good idea to have more specific error messages here. Otherwise it's hard to understand which part of the output is empty. (n.b. `fail` will not report the line it failed on, so calling it twice with the same string should be avoided if possible).

**Review by gzm0 — CHANGES_REQUESTED** [review]
Mostly minor stuff, only larger point: is --outputDirectory good enough without short version. @sjrd might also have an opinion on this.

**Comment by sjrd** [comment]
> Mostly minor stuff, only larger point: is --outputDirectory good enough without short version. @sjrd might also have an opinion on this.

Indeed, this option should have a short name. If you can't find anything meaningful, use something meaningless like `-z`. ;)

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (clutroth) [reply src/main/scala/org/scalajs/cli/Scalajsld.scala]: I missed this doc. I missed this ugly line break, too.

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (clutroth) [reply src/main/scala/org/scalajs/cli/Scalajsld.scala]: reverted

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (clutroth) [reply src/main/scala/org/scalajs/cli/Scalajsld.scala]: That was my first approach, but `DeprecatedFileOutputSilencer.link` is object's method that is deprecated. `DeprecatedFileOutputSilencer().link` is legal class's method. I have to invoke class's method.

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (clutroth) [reply src/main/scala/org/scalajs/cli/Scalajsld.scala]: Honestly, I'm not familiar with ScalaJS, specially it's internals. I understand `outputPatterns` form CLI perspective, as target filenames template.

Maybe """File pattern used to create output files, and relative URI pattern, By default "%s.js""""? I feel like I'm not having right context. @gzm0

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: Braces are required if the conditional block spans multiple lines:

```suggestion
        if (c.output.isDefined) {
          reportWarning("using a single file as output (--output) is deprecated since Scala.js 1.3.0." +
            " Use --outputDirectory instead.")
        }
```

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]:
```suggestion
        .flatMap { irFiles =>
```

Space before brace :)

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: Always use parenthesis on Java APIs.

```suggestion
              linker.link(irFiles, moduleInitializers, PathOutputDirectory(outputDir.toPath()), logger)
```

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: Does this not work:

```suggestion
              (DeprecatedLinkerAPI: DeprecatedLinkerAPI).link(linker, irFiles.toList, moduleInitializers, jsFile, logger)
```

Then you won't need the additional `apply` method on `object DeprecatedLinkerAPI`.

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: Throw an `AssertionError`? If we get here: The previous code has failed. So it's not a user error anymore, but a bug in the code.

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: The indentation here is not following SJS style: https://github.com/scala-js/scala-js/blob/master/CODINGSTYLE.md#indentation

Either binpack it (as many args per line as possible, 4 spaces indent), or one per line, 2 spaces indented.

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [reply src/main/scala/org/scalajs/cli/Scalajsld.scala]: How about: "Pattern for JS file names (default: `%s.js`). Expects a printf-style pattern with a single placeholder for the module ID. A typical use case is changing the file extension, e.g. `%.mjs` for Node.js modules."?

**Review by gzm0 — CHANGES_REQUESTED** [review]
Sorry this took so long. I'm struggling to find time for Scala.js.

I suggested a help text for `jsFilePattern`. Otherwise only codingstyle nits.

If all of this is addressed, feel free to squash everything into a single commit right away.

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (clutroth) [reply src/main/scala/org/scalajs/cli/Scalajsld.scala]: I'm not sure, if it's this case. It'd be right for function invocation, but no declaration. I applied rule `In general, indentation is 2 spaces, except continuation lines, which are indented 4 spaces.` following example `org.scalajs.linker.interface.Linker#link`

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: Use backticks for the examples? (like in my comment, but the markdown version of it)?

**Comment by clutroth** [comment]
@gzm0 could we proceed?

File: src/main/scala/org/scalajs/cli/Scalajsld.scala
Comment (gzm0) [inline src/main/scala/org/scalajs/cli/Scalajsld.scala]: `--outputDir`.

**Review by gzm0 — CHANGES_REQUESTED** [review]
Sorry, this slipped through the cracks. Thank you for reminding me. Only minor things, really.

**Review by gzm0 — COMMENTED** [review]
LGTM. Please squash all the commits and then we can merge this (unless @sjrd has comments / objections).

**Comment by sjrd** [comment]
No objection. ;) Looks good to me.

**Review by gzm0 — APPROVED** [review]

**Pull request merged**
